File: 2019/src/main/kotlin/day08/day8.kt
package day08

import java.io.File

fun parseInput(input: String): List<List<Int>> = input.map { it - '0' }.chunked(6 * 25)

fun evaluatorOne(layers: List<List<Int>>): Int {
    val layer = layers.minBy { layer -> layer.count { it == 0 } }
    val ones = layer.count { it == 1 }
    val twos = layer.count { it == 2 }
    return ones * twos
}

fun evaluatorTwo(layers: List<List<Int>>): String {
    val img = CharArray(6 * 25) { ' ' }

    for (layer in layers.asReversed()) {
        for (i in img.indices) {
            img[i] = when (layer[i]) {
                0 -> ' '
                1 -> '#'
                else -> img[i]
            }
        }
    }

    return img.toList().chunked(25).joinToString("\n") { it.joinToString("") }
}

fun readLinesFromFile(filePath: String): Result<List<String>> =
    runCatching { File(filePath).readLines() }

fun main() {
    readLinesFromFile("src/main/resources/day08.txt")
        .onSuccess {
            val layers = parseInput(it.first())
            println("Part One: ${evaluatorOne(layers)}")
            println("Part Two:\n${evaluatorTwo(layers)}")
        }
        .onFailure {
            println("Error reading file: ${it.message}")
        }
}

File: 2019/src/main/scala/day08/day8.scala
package day08

import scala.util.{Try, Success, Failure, Using}
import scala.io.Source

def parseInput(input: String) = input.map(_.asDigit).grouped(6 * 25).toList

def evaluatorOne(layers: List[IndexedSeq[Int]]): Int = {
    val frequencies = layers.minBy(_.count(_ == 0)).groupMapReduce(identity)(_ => 1)(_ + _)
    return frequencies(1) * frequencies(2)
}

def evaluatorTwo(layers: List[IndexedSeq[Int]]): String = {
    val img = layers.transpose.map(pixelStack => {
        pixelStack.find(_ != 2) match {
            case Some(1) => '#'
            case _       => ' '
        }
    })

    return img.grouped(25).map(_.mkString).mkString("\n")
}

def readLinesFromFile(filePath: String): Try[List[String]] =
    Using(Source.fromResource(filePath))(_.getLines().toList)

def hello(): Unit = {
    readLinesFromFile("day08.txt") match {
        case Success(lines) => {
            val layers = parseInput(lines.head)
            println(s"Part One: ${evaluatorOne(layers)}")
            println(s"Part Two:\n${evaluatorTwo(layers)}")
        }
        case Failure(exception) => {
            println(s"Error reading file: ${exception.getMessage}")
        }
    }
}